### PR TITLE
fmpy: init at 0.3.1

### PR DESCRIPTION
**Pull request by balodja**
###### Motivation for this change
Useful library for dealing with Functional Mock-up Units (see [FMI](https://fmi-standard.org/)) in Python. This is continuation of work in #122828 and #131504

###### Things done
- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -1124,6 +1124,12 @@
     email = "sivaraman.balaji@gmail.com";
     name = "Balaji Sivaraman";
   };
+  balodja = {
+    email = "balodja@gmail.com";
+    github = "balodja";
+    githubId = 294444;
+    name = "Vladimir Korolev";
+  };
```
Comment (jonringer) [inline maintainers/maintainer-list.nix:1132]: Please make this into a separate commit:

```
maintainers: add balodja
```

File: pkgs/development/python-modules/fmpy/default.nix [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      cmake --build fmpy/fmucontainer/${platform} --config Release
+
+      # The reproduction of build_remoting.py
+      cmake -S remoting -B remoting/linux64 -D RPCLIB=${rpclib}
+      cmake --build remoting/linux64 --config Release
+    '';
+
+  # Some of tests are supposed to be failing in upstream, so we don't use
+  # them for package verification at the moment.
+  doCheck = false;
```
Comment (jonringer) [inline pkgs/development/python-modules/fmpy/default.nix:132]: If tests are not available, then please use `pythonImportsCheck` to import the most important modules. This isn't as good as unit tests, but will usually give a good indication of run-time errors. Please see [pythonImportsCheck documentation](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pythonimportscheck) for more information.

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (balodja) [reply maintainers/maintainer-list.nix:1132]: Done!

File: pkgs/development/python-modules/fmpy/default.nix (hunk 2 above)
Comment (balodja) [reply pkgs/development/python-modules/fmpy/default.nix:131]: Done! That's a good suggestion. Once the tests will be ready, I will fix that.

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 131741` at e48f014a run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/eff093d323577afb95687333b48094a7">python38Packages.fmpy</a></li>
    <li><a href="https://gist.github.com/63c5fa7113af4373e3b770d51b5034d4">python39Packages.fmpy</a></li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

File: pkgs/development/python-modules/fmpy/default.nix [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+let
+  # Auxiliary library, maybe not extremely useful by itself.
+  rpclib = stdenv.mkDerivation {
+    pname = "rpclib";
+    version = "2.3.0";
+
+    src = fetchFromGitHub {
+      owner = "rpclib";
+      repo = "rpclib";
+      rev = "v2.3.0";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/fmpy/default.nix:31]:
```suggestion
  rpclib = stdenv.mkDerivation rec {
    pname = "rpclib";
    version = "2.3.0";

    src = fetchFromGitHub {
      owner = "rpclib";
      repo = "rpclib";
      rev = "v${version}";
```

File: pkgs/development/python-modules/fmpy/default.nix [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+, dask
+, requests
+, matplotlib
+, pyqt5
+, pyqtgraph
+, notebook
+, plotly
+}:
+let
+  # Auxiliary library, maybe not extremely useful by itself.
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/fmpy/default.nix:23]: I think we should move them anyway to all-packages just incase they are used somewhere else down the line.

File: pkgs/development/python-modules/fmpy/default.nix [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+  };
+
+  cvode = stdenv.mkDerivation {
+    pname = "cvode";
+    version = "5.3.0";
+
+    src = fetchFromGitHub {
+      owner = "LLNL";
+      repo = "sundials";
+      rev = "v5.3.0";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/fmpy/default.nix:45]:
```suggestion
  cvode = stdenv.mkDerivation rec {
    pname = "cvode";
    version = "5.3.0";

    src = fetchFromGitHub {
      owner = "LLNL";
      repo = "sundials";
      rev = "v${version}";
```

File: pkgs/development/python-modules/fmpy/default.nix [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+buildPythonPackage {
+  pname = "FMPy";
+  version = "0.3.1";
+  disabled = pythonOlder "3.5";
+
+  src = fetchFromGitHub {
+    owner = "CATIA-Systems";
+    repo = "FMPy";
+    rev = "v0.3.1";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/fmpy/default.nix:71]:
```suggestion
buildPythonPackage rec {
  pname = "FMPy";
  version = "0.3.1";
  disabled = pythonOlder "3.5";

  src = fetchFromGitHub {
    owner = "CATIA-Systems";
    repo = "FMPy";
    rev = "v${version}";
```

File: pkgs/development/python-modules/fmpy/default.nix [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    sha256 = "0ig08323id84nd9y272hyccky33my2nlhl44m41fxnkrg6xq7z3m";
+  };
+
+  nativeBuildInputs = [ cmake pyqt5 ];
+
+  propagatedBuildInputs = [
+    cmake
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/fmpy/default.nix:78]: You are sure about this?

File: pkgs/development/python-modules/fmpy/default.nix [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+      cmake -S fmpy/logging -B fmpy/logging/build
+      cmake --build fmpy/logging/build --config Release
+
+      # The reproduction of build_fmucontainer.py:
+      cmake -S fmpy/fmucontainer -B fmpy/fmucontainer/${platform}
+      cmake --build fmpy/fmucontainer/${platform} --config Release
+
+      # The reproduction of build_remoting.py
+      cmake -S remoting -B remoting/linux64 -D RPCLIB=${rpclib}
+      cmake --build remoting/linux64 --config Release
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/fmpy/default.nix:126]: This is missing cmakeFlags.

File: pkgs/development/python-modules/fmpy/default.nix [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    "fmpy.ssp"
+    "fmpy.sundials"
+  ];
+
+  meta = with lib; {
+    description = "A free Python library to simulate Functional Mock-up Units (FMUs)";
+    homepage = "https://github.com/CATIA-Systems/FMPy";
+    license = with licenses; [ gpl3Only ];
+    maintainers = with maintainers; [ balodja ];
+    platforms = platforms.unix;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/fmpy/default.nix:151]:
```suggestion
```
that is the default for buildPythonPackage.

File: pkgs/development/python-modules/fmpy/default.nix (hunk 2 above)
Comment (risicle) [reply pkgs/development/python-modules/fmpy/default.nix:122]: You can also omit certain tests quite easily if that's better.

**Comment by risicle** [comment]
Boring thing, but python packages' commit messages should generally be e.g.

```
python3Packages.fmpy: init at 0.3.1
```

**Comment by risicle** [comment]
Hmm. Fails to build on macos 10.15:

```
-- Looking for pthread.h
-- Looking for pthread.h - not found
CMake Error at /nix/store/vl9s7y2945kkb765r4hyznzss0yxy1wr-cmake-3.19.7/share/cmake-3.19/Modules/FindPackageHandleStandardArgs.cmake:218 (message):
  Could NOT find Threads (missing: Threads_FOUND)
Call Stack (most recent call first):
  /nix/store/vl9s7y2945kkb765r4hyznzss0yxy1wr-cmake-3.19.7/share/cmake-3.19/Modules/FindPackageHandleStandardArgs.cmake:582 (_FPHSA_FAILURE_MESSAGE)
  /nix/store/vl9s7y2945kkb765r4hyznzss0yxy1wr-cmake-3.19.7/share/cmake-3.19/Modules/FindThreads.cmake:234 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  CMakeLists.txt:107 (find_package)


-- Configuring incomplete, errors occurred!
```

File: pkgs/development/python-modules/fmpy/default.nix (hunk 2 above)
Comment (balodja) [reply pkgs/development/python-modules/fmpy/default.nix:100]: Thanks for suggestion. Failing tests are not the only problem. Most of the tests download FMUs from the internet with their own code. I'll make a PR to FMPy to make tests more Nix-friendly before switching them on.

File: pkgs/development/python-modules/fmpy/default.nix (hunk 3 above)
Comment (balodja) [reply pkgs/development/python-modules/fmpy/default.nix]: Thank you, accepted!

File: pkgs/development/python-modules/fmpy/default.nix (hunk 4 above)
Comment (balodja) [reply pkgs/development/python-modules/fmpy/default.nix]: Done. I had to add more options to `sundials` package to make it more compliant with the FMPy. Hope that's fine.

File: pkgs/development/python-modules/fmpy/default.nix (hunk 5 above)
Comment (balodja) [reply pkgs/development/python-modules/fmpy/default.nix]: Done.

File: pkgs/development/python-modules/fmpy/default.nix (hunk 7 above)
Comment (balodja) [reply pkgs/development/python-modules/fmpy/default.nix]: Double checked. You are right, I'm not sure. The FMPy generates `CMakeLists.txt`, but it runs `cmake` by itself only in tests, not in regular case.

File: pkgs/development/python-modules/fmpy/default.nix (hunk 8 above)
Comment (balodja) [reply pkgs/development/python-modules/fmpy/default.nix]: Added cmakeFlags in those invocations. With a bit of a hack, but it should be OK.

File: pkgs/development/python-modules/fmpy/default.nix (hunk 9 above)
Comment (balodja) [reply pkgs/development/python-modules/fmpy/default.nix]: OK, thank you!

**Comment by balodja** [comment]
> python3Packages.fmpy

Thanks! Corrected.

File: pkgs/development/python-modules/fmpy/default.nix [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+, matplotlib
+, pyqt5
+, pyqtgraph
+, notebook
+, plotly
+}:
+
+# FMPy respects ILP64 only
+assert stdenv.targetPlatform.is64bit;
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/fmpy/default.nix:27]: Please use platform, broken instead.

Comment (balodja) [reply pkgs/development/python-modules/fmpy/default.nix]: Sorry, assert was really a bad choice. Corrected.

**Comment by doronbehar** [comment]
This seems good overall, @balodja would you like to rebase and fix merge conflicts?

**Comment by SuperSandro2000** [comment]
This PR is pretty stale at this point and closing because of that.

**Comment by tmplt** [comment]
The work continues in the PR linked above.
